src/linear_algebra/dual.lean

```diff
@@ -398,3 +407,145 @@ begin
 end
 
 end dual_pair
+
+namespace submodule
+
+universes u v w
+
+variables {R : Type u} {M : Type v} [comm_ring R] [add_comm_group M] [module R M]
+variable {W : submodule R M}
+
+/-- The `dual_restrict` of a submodule `W` of `M` is the linear map from the
+  dual of `M` to the dual of `W` such that the domain of each linear map is
+  restricted to `W`. -/
+def dual_restrict (W : submodule R M) :
+  module.dual R M →ₗ[R] module.dual R W :=
+linear_map.dom_restrict' W
+
+@[simp] lemma dual_restrict_apply
+  (W : submodule R M) (φ : module.dual R M) (x : W) :
+  W.dual_restrict φ x = φ x.1 := rfl
+
+/-- The `dual_annihilator` of a submodule `W` is the set of linear maps `φ` such
+  that `φ w = 0` for all `w ∈ W`. -/
+def dual_annihilator {R : Type u} {M : Type v} [comm_ring R] [add_comm_group M]
+  [module R M] (W : submodule R M) : submodule R $ module.dual R M :=
+W.dual_restrict.ker
+
+@[simp] lemma mem_dual_annihilator (φ : module.dual R M) :
+  φ ∈ W.dual_annihilator ↔ ∀ w ∈ W, φ w = 0 :=
+begin
+  refine linear_map.mem_ker.trans _,
+  simp_rw [linear_map.ext_iff, dual_restrict_apply],
+  exact ⟨λ h w hw, h ⟨w, hw⟩, λ h w, h w.1 w.2⟩
+end
+
+lemma dual_restrict_ker_eq_dual_annihilator
+  (W : submodule R M) : W.dual_restrict.ker = W.dual_annihilator :=
+rfl
+
+end submodule
+
+namespace subspace
+
+open submodule linear_map
+
+universes u v w
+
+-- We work in vector spaces because `exists_is_compl` only hold for vector spaces
+variables {K : Type u} {V : Type v} [field K] [add_comm_group V] [vector_space K V]
+
+/-- Given a subspace `W` of `V` and an element of its dual `φ`, `dual_lift W φ` is
+  the natural extenstion of `φ` to an element of the dual of `V`. -/
```

Comment:
Did I understand this correctly?
```suggestion
the natural extension of `φ` to an element of the dual of `V`.
That is, `dual_lift W φ` sends `x ∈ W` to `φ x` and `x` orthogonal to `W` to `0`.
-/
```

Reply:
I'm not sure the word "orthogonal" is correct since a subspace can be a complement without being orthogonal. Perhaps
``` `dual_lift W φ` sends `x ∈ W` to `φ x` and `x` in the complement of `W` to `0`. ```